paymentsheet/src/main/java/com/stripe/android/paymentelement/confirmation/gpay/GooglePayConfirmationDefinition.kt

```diff
@@ -0,0 +1,179 @@
+        return confirmationOption as? GooglePayConfirmationOption
+    }
+
+    override suspend fun action(
+        confirmationOption: GooglePayConfirmationOption,
+        intent: StripeIntent
+    ): ConfirmationDefinition.Action<Unit> {
+        if (
+            confirmationOption.config.merchantCurrencyCode == null &&
+            !confirmationOption.initializationMode.isProcessingPayment
+        ) {
+            val message = "GooglePayConfig.currencyCode is required in order to use " +
+                "Google Pay when processing a Setup Intent"
+
+            userFacingLogger?.logWarningWithoutPii(message)
+
+            return ConfirmationDefinition.Action.Fail(
+                cause = IllegalStateException(message),
+                message = R.string.stripe_something_went_wrong.resolvableString,
+                errorType = ConfirmationHandler.Result.Failed.ErrorType.MerchantIntegration,
+            )
+        }
+
+        return ConfirmationDefinition.Action.Launch(
+            launcherArguments = Unit,
+            receivesResultInProcess = true,
+            deferredIntentConfirmationType = null,
+        )
+    }
+
+    override fun createLauncher(
+        activityResultCaller: ActivityResultCaller,
+        onResult: (GooglePayPaymentMethodLauncher.Result) -> Unit
+    ): ActivityResultLauncher<GooglePayPaymentMethodLauncherContractV2.Args> {
+        return activityResultCaller.registerForActivityResult(
+            GooglePayPaymentMethodLauncherContractV2(),
+            onResult,
+        )
+    }
+
+    override fun launch(
+        launcher: ActivityResultLauncher<GooglePayPaymentMethodLauncherContractV2.Args>,
+        arguments: Unit,
+        confirmationOption: GooglePayConfirmationOption,
+        intent: StripeIntent,
+    ) {
+        val config = confirmationOption.config
+        val googlePayLauncher = createGooglePayLauncher(
+            factory = googlePayPaymentMethodLauncherFactory,
+            activityLauncher = launcher,
+            config = confirmationOption.config,
+        )
+
+        googlePayLauncher.present(
+            currencyCode = intent.asPaymentIntent()?.currency
+                ?: config.merchantCurrencyCode.orEmpty(),
+            amount = when (intent) {
+                is PaymentIntent -> intent.amount ?: 0L
+                is SetupIntent -> config.customAmount ?: 0L
+            },
+            transactionId = intent.id,
+            label = config.customLabel,
+        )
+    }
+
+    override fun toResult(
+        confirmationOption: GooglePayConfirmationOption,
+        deferredIntentConfirmationType: DeferredIntentConfirmationType?,
+        intent: StripeIntent,
+        result: GooglePayPaymentMethodLauncher.Result,
+    ): ConfirmationDefinition.Result {
+        return when (result) {
+            is GooglePayPaymentMethodLauncher.Result.Completed -> {
+                val nextConfirmationOption = PaymentMethodConfirmationOption.Saved(
```

Comment:
I'm not sure this is perfect. But at a minimum, so much better than modeling this as a `PaymentSelection`!

Reply:
Agreed! I'd like to come up with a solution in which we don't model from one option to another. It is hard though because we have quite a few options that convert from one option to another or modify an option. Something we should think about after the architecture changes.

Reply:
Thinking more about this. It does kinda seem like the job of this definition to do this. So it feels right.